cmd/api/main.go

```diff
@@ -419,3 +396,77 @@ func createChainIDStack(
 		},
 	}, nil
 }
+
+func configureTelemetry(
+	nodeID string,
+	dirPath string,
+	chainStacks map[tableland.ChainID]chains.ChainStack,
+	config TelemetryPublisherConfig,
+) (moduleClose, error) {
+	// Wiring
+	metricsDatabaseURL := fmt.Sprintf(
+		"file://%s?_busy_timeout=5000&_foreign_keys=on&_journal_mode=WAL",
+		path.Join(dirPath, "metrics.db"),
+	)
+
+	metricsStore, err := storage.New(metricsDatabaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("creating metrics store: %s", err)
+	}
+
+	var metricsPublisher *publisher.Publisher
+	if config.Enabled {
+		exporter, err := publisher.NewHTTPExporter(
+			config.MetricsHubURL,
+			config.MetricsHubAPIKey,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("creating metrics http exporter: %s", err)
+		}
+		publishingInterval, err := time.ParseDuration(config.PublishingInterval)
+		if err != nil {
+			return nil, fmt.Errorf("parsing publishing interval: %s", err)
+		}
+		metricsPublisher = publisher.NewPublisher(metricsStore, exporter, nodeID, publishingInterval)
+		metricsPublisher.Start()
+	}
+	telemetry.SetMetricStore(metricsStore)
+
+	// Collect binary information.
+	ctx, cls := context.WithTimeout(context.Background(), time.Second)
+	defer cls()
+	if err := telemetry.Collect(ctx, buildinfo.GetSummary()); err != nil {
+		return nil, fmt.Errorf("collect git summary: %s", err)
+	}
+
+	collectFrequency, err := time.ParseDuration(config.ChainStackCollectFrequency)
+	if err != nil {
+		return nil, fmt.Errorf("invalid chain stack collect frequency configuration: %s", err)
+	}
+	ctxChainStackCollector, clsChainStackCollector := context.WithCancel(context.Background())
+	chainStackCollectorClosed := make(chan struct{})
+	cc, err := chainscollector.New(chainStacks, collectFrequency)
+	if err != nil {
+		clsChainStackCollector()
+		return nil, fmt.Errorf("configure chains collector: %s", err)
+	}
+	go func() {
+		cc.Start(ctxChainStackCollector)
+		close(chainStackCollectorClosed)
+	}()
+
+	// Module closing function
+	return func(ctx context.Context) error {
+		clsChainStackCollector()
+		<-chainStackCollectorClosed
+
+		if err := metricsStore.Close(); err != nil {
+			return fmt.Errorf("closing metrics store: %s", err)
+		}
+		if config.Enabled {
+			metricsPublisher.Close()
+		}
+
+		return nil
```

Comment:
This is the closing function returned to `main`. This allows us to clearly express all the needed work to close the telemetry module properly. The logic here isn't mixed with other wires of main.

Also, `main` will simply have to execute this returned function without knowing what it does. It simply knows that it will do the right thing.